### PR TITLE
[WiP] Lodash 3 refactors

### PR DESCRIPTION
**Pull request by megawac**
Please review before further work continues...

~~Was working on this over the past 40 minutes or so, let me know if I should continue.~~

Note to self
  - refactor query/finders/basic `find` and `findOne` in a separate PR

### CODE REVIEW

**Comment by Globegitter** [comment]
:+1: I would love to see that. One thing I would suggest though is requiring `lodash` modules directly, at least where it makes sense.
So for example you can now do 
```js
var isObject = require('lodash/Lang/isObject');
var reduce = require('lodash/Collection/reduce');
```
which can improve the performance.

**Comment by tjwebb** [comment]
+1

File: lib/waterline/utils/normalize.js [hunk 1]
```diff
@@ -192,11 +192,8 @@ var normalize = module.exports = {
     // empty object without running the query because nothing will match anyway. Let's return
     // false from here so the query knows to exit out.
     if(criteria.where) {
-      var falsy = false;
-      Object.keys(criteria.where).forEach(function(key) {
-        if(Array.isArray(criteria.where[key]) && criteria.where[key].length === 0) {
-          falsy = true;
-        }
+      var falsy = _.any(criteria.where, function(prop) {
+        return _.isArray(prop) && prop.length === 0;
```
Comment (megawac) [inline lib/waterline/utils/normalize.js]: This can probably just be `_.any(criteria.where, _.isEmpty)`

**Comment by megawac** [comment]
> One thing I would suggest though is requiring lodash modules directly, at least where it makes sense. [...] which can improve performance

In this case we're using so much of lodash that I doubt that will make much sense. It would more likely degrade performance requiring each module piece by piece whereas this is using the prebuilt `index.js`

Can people make sure I'm not breaking anything before this PR gets too out of hand.

**Comment by Globegitter** [comment]
@megawac Not so sure about that: https://github.com/babel/babel switched from requiring `_` all over the place to modules directly and `requiring` time dropped quite significantly. From about 2s to below 1s. There may have been some more performance improvements, but lodash was certainly one of the biggest factors.

And yeah will test in my local project asap.

**Comment by listepo** [comment]
+1

**Comment by megawac** [comment]
> :+1: I would love to see that. One thing I would suggest though is requiring lodash modules directly

@Globegitter made a tool a little while ago pretty much based on your comment https://github.com/megawac/lodash-modularize

**Comment by particlebanana** [comment]
There is just too much to do at once. As pieces get refactored we can move to more lodash features but doing it all in one pass doesn't seem worth it and would be too much to try and verify the pr against.

**Pull request closed without merging**